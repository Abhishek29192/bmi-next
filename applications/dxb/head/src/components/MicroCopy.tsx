import { graphql } from "gatsby";
import { MicroCopyValues } from "@bmi/microcopies";

export type Data = {
  key: string;
  value: string;
};

export type GetMicroCopy = (
  path: MicroCopyValues,
  variables?: Record<string, string>
) => string;

export const fallbackGetMicroCopy: GetMicroCopy = (path) => `MC: ${path}`;

export const generateGetMicroCopy = (microCopy?: Data[] | null) => {
  if (!microCopy) {
    return fallbackGetMicroCopy;
  }

  const replaceVariables: GetMicroCopy = (copy = "", variables = {}) =>
    Object.entries(variables).reduce(
      (acc, [key, value]) => acc.replace(`{{${key}}}`, value),
      copy
    );

  const getMicroCopy: GetMicroCopy = (path, variables) => {
    const copy = microCopy.find(({ key }) => key === path)?.value;
    return (
      replaceVariables(copy as MicroCopyValues, variables) || `MC: ${path}`
    );
  };

  return getMicroCopy;
};

export const query = graphql`
  fragment MicroCopyFragment on ContentfulMicroCopy {
    key
    value
  }
`;
