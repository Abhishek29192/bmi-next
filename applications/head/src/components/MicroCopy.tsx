import { graphql } from "gatsby";

export type Data = {
  key: string;
  value: string;
};

export type GetMicroCopy = (path: string) => string;

export const fallbackGetMicroCopy: GetMicroCopy = (path) => `MC: ${path}`;

export const generateGetMicroCopy = (microCopy?: Data[]) => {
  if (!microCopy) {
    return fallbackGetMicroCopy;
  }

  const getMicroCopy: GetMicroCopy = (path) => microCopy[path] || `MC: ${path}`;

  return getMicroCopy;
};

export const query = graphql`
  fragment MicroCopyFragment on ContentfulMicroCopy {
    key
    value
  }
`;
