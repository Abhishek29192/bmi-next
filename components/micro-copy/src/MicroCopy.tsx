import React, { createContext, useContext } from "react";
import escapeStringRegexp from "escape-string-regexp";

export const MicroCopyContext = createContext<Record<string, string>>({});

export const getMicroCopy = (
  values: Record<string, string> = {},
  path: string,
  placeholders: Record<string, string> = {},
  prefixMC: boolean = false
): string =>
  Object.entries(placeholders).reduce((carry, [key, value]) => {
    const toReplace = `{{${key}}}`;
    return carry.replace(new RegExp(escapeStringRegexp(toReplace), "g"), value);
  }, values[path] || (prefixMC ? `MC:${path}` : path));

type Props = {
  path: string;
  placeholders?: Record<string, string>;
};

const MicroCopy = ({ path, placeholders = {} }: Props) => {
  const values = useContext(MicroCopyContext);
  const value: string = getMicroCopy(values, path, placeholders);

  return <>{value}</>;
};

type ProviderProps = {
  values: Record<string, string>;
  children: React.ReactNode;
};

const MicroCopyProvider = ({ values, children }: ProviderProps) => {
  return (
    <MicroCopyContext.Provider value={values}>
      {children}
    </MicroCopyContext.Provider>
  );
};

MicroCopy.Provider = MicroCopyProvider;

export default MicroCopy;
