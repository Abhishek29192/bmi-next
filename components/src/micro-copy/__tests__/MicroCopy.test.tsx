import React from "react";
import { render } from "@testing-library/react";
import MicroCopy, { getMicroCopy } from "../MicroCopy";

const ProviderMock = ({ children }: { children: React.ReactNode }) => {
  return (
    <MicroCopy.Provider
      values={{
        first: "Lorem firstKey",
        second: `
          Lorem {{firstKey}};
          Another occurrence: {{firstKey}};
          A second key: {{secondKey}};`
      }}
    >
      {children}
    </MicroCopy.Provider>
  );
};

describe("MicroCopy tests", () => {
  describe("getMicroCopy function tests", () => {
    describe("When Microcopy is found", () => {
      describe("And function is called with prefixMC = true", () => {
        it("renders correctly without MC prefix", () => {
          const result = getMicroCopy(
            { someKey: "someValue" },
            "someKey",
            {},
            true
          );
          expect(result).toEqual("someValue");
        });
      });
      describe("And function is called with prefixMC = false", () => {
        it("renders correctly without MC prefix", () => {
          const result = getMicroCopy(
            { someKey: "someValue" },
            "someKey",
            {},
            false
          );
          expect(result).toEqual("someValue");
        });
      });
    });
    describe("When Microcopy is NOT found", () => {
      describe("And function is called with prefixMC = true", () => {
        it("renders correctly with MC prefix", () => {
          const result = getMicroCopy(
            { someKey: "someValue" },
            "test",
            {},
            true
          );
          expect(result).toEqual("MC:test");
        });

        it("renders correctly when values is undefined", () => {
          const result = getMicroCopy(undefined, "test", {}, true);
          expect(result).toEqual("MC:test");
        });

        it("renders correctly when placeholders is undefined", () => {
          const result = getMicroCopy(undefined, "test", undefined, true);
          expect(result).toEqual("MC:test");
        });
      });

      describe("And function is called with prefixMC = false", () => {
        it("renders correctly without MC prefix", () => {
          const result = getMicroCopy(
            { someKey: "someValue" },
            "test",
            {},
            false
          );
          expect(result).toEqual("test");
        });
      });
    });
  });

  describe("MicroCopy component", () => {
    it("renders correctly without replacement", () => {
      const { container } = render(
        <>
          <ProviderMock>
            Without placeholders (but they are passed):{" "}
            <MicroCopy
              path="first"
              placeholders={{
                firstKey: `This shouldn't be showing because the key usage isn't wrapped with {{}}`
              }}
            />
            With placeholders (but they are not passed):{" "}
            <MicroCopy path="second" />
          </ProviderMock>
          <MicroCopy path="not.existing" />
        </>
      );
      expect(container).toMatchSnapshot();
    });

    it("replaces all occurrences", () => {
      const { container } = render(
        <ProviderMock>
          <MicroCopy
            path="second"
            placeholders={{
              firstKey: "firstKey replaced",
              secondKey: "second key replaced"
            }}
          />
        </ProviderMock>
      );
      expect(container).toMatchSnapshot();
    });

    it("does not replace placeholders unless passed", () => {
      const { container } = render(
        <ProviderMock>
          Only firstKey should be replaced:{" "}
          <MicroCopy
            path="second"
            placeholders={{
              firstKey: "firstKey replaced"
            }}
          />
        </ProviderMock>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
