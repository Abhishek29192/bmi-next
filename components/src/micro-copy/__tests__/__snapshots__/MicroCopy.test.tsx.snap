// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MicroCopy tests MicroCopy component does not replace placeholders unless passed 1`] = `
<div>
  Only firstKey should be replaced:
   
  
          Lorem firstKey replaced;
          Another occurrence: firstKey replaced;
          A second key: {{secondKey}};
</div>
`;

exports[`MicroCopy tests MicroCopy component renders correctly without replacement 1`] = `
<div>
  Without placeholders (but they are passed):
   
  Lorem firstKey
  With placeholders (but they are not passed):
   
  
          Lorem {{firstKey}};
          Another occurrence: {{firstKey}};
          A second key: {{secondKey}};
  not.existing
</div>
`;

exports[`MicroCopy tests MicroCopy component replaces all occurrences 1`] = `
<div>
  
          Lorem firstKey replaced;
          Another occurrence: firstKey replaced;
          A second key: second key replaced;
</div>
`;
